giskard/scanner/llm/llm_prompt_injection_detector.py

```diff
@@ -30,89 +35,58 @@ class LLMPromptInjectionDetector(Detector):
-                if column_types[feature] == "text":
-                    prompt_dataset.df[feature] = prompt.content
-
-            prediction = model.predict(prompt_dataset).prediction
-            if prediction.shape[0] > 1:
-                raise ValueError("The prediction is expected to be 1D.")
-            prediction = prediction[0]
-
-            failed = evaluate(prediction=prediction, prompt=prompt)
-
-            if prompt.group not in results.keys():
-                results[prompt.group] = {"prompt_name": [], "failed": [], "input_prompt": [], "prediction": []}
-
-            results[prompt.group]["failed"].append(failed)
-            results[prompt.group]["input_prompt"].append(prompt)
-            results[prompt.group]["prediction"].append(prediction)
-        return results
-
     def run(self, model: BaseModel, dataset: Dataset) -> Sequence[Issue]:
-        logger.info(
-            f"Running the {Style.RESET_ALL}{Fore.LIGHTBLUE_EX}{self.__class__.__name__}{Style.RESET_ALL} Detector."
-        )
-
-        # even-though this detector doesn't rely on a dataset, it's still needed to get the features and column_types
-        features = model.meta.feature_names or list(dataset.df.columns.drop(dataset.target, errors="ignore"))
-        column_types = dataset.column_types
-
-        prompts = get_all_prompts()
+        generator = PromptInjectionDataLoader(num_samples=self.num_samples)
+        dataset = generator.load_dataset(dataset.column_types)
+        meta_df = generator.all_meta_df
 
+        evaluator = StringMatcher()
         issues = []
-
-        results = self.evaluate_and_group(model, dataset, prompts, features, column_types)
-
-        for group in results.keys():
-            failed_examples = {}
-            cols = []
-            prompts_names = [prompt.name for prompt in results[group]["input_prompt"]]
-            prompts_contents = [prompt.content for prompt in results[group]["input_prompt"]]
-            if all(prompts_names):
-                failed_examples.update({"Prompt Name": prompts_names})
-                cols = ["Prompt Name"]
-
-            failed_examples.update({"Input Prompt": prompts_contents, "Prediction": results[group]["prediction"]})
-            cols += ["Input Prompt", "Prediction"]
-
-            failed_examples_df = pd.DataFrame(failed_examples)[cols]
-            failed_examples_df = failed_examples_df.filter(
-                items=[i for i, v in enumerate(results[group]["failed"]) if v != 0], axis=0
-            )
-
-            failed = sum(results[group]["failed"])
-            if failed == 0:
+        for group in set(generator.groups_mapping):
+            group_idx = meta_df.index[meta_df["group_mapping"] == group].tolist()
+            group_dataset = dataset.slice(group_slice(group_idx=group_idx))
+            group_meta_df = meta_df.iloc[group_idx]
+            evaluation_results = evaluator.evaluate(model, group_dataset, group_meta_df.to_dict("records"))
```

Comment:
Can we separate a bit the logic of the grouping? It‘s very confusing here.

If you have to define groups of prompts, maybe let’s do that at the generator level? E.g. generator returning a list of datasets to test.

```py
datasets = loader.load_datasets_by_group()
for eval_dataset in datasets:
    eval_result = evaluator.evaluate(model, eval_dataset)
```

Reply:
Also, I feel that this could be done with a left join on the data frame level instead of manually checking indexes.

Reply:
eventually done it in this PR